### PR TITLE
V2 default product handling

### PR DESCRIPTION
**Pull request by maira-samtek**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,282 +1,584 @@
...
 1. Gets all current findings from Security Hub, filtered by the `Severity` of the findings (default is `CRITICAL,HIGH,MEDIUM`)
 
+  
+
 2. Creates Jira issues for findings that do not already have a Jira issue
 
-- **To avoid creating duplicate issues**, the search criteria use the custom label configuration as well as default configuration for three identifying search labels: region, AWS account ID, and "security-hub."
+  
+
+-  **To avoid duplicate issues**, the search criteria use default settings for three primary identifiers: region, AWS account ID, and "security-hub," along with any custom label configurations provided (Note that modifications or deletions of labels may lead to duplicate issue creation). Additionally, a comparison of the Jira issue description and finding title is made against the findings list. If there is a title match, the decision is based on the Resources list.
```
Comment (yonassrobi) [inline README.md]: Would it make sense to rename "security-hub," to "Security Hub title" or something a bit more descriptive?

Comment (yonassrobi) [reply README.md]: Can you provide an example of "custom label configuration" here? 

It's good that you included the "Note" about modifying custom labels here. Can this be bolded/italicized and on a different line to make sure it's clearly visible?

File: README.md [hunk 2]
```diff
@@ -1,282 +1,584 @@
...
+- For example, if a Jira ticket T1 already exists for a finding with title T1, and a new finding appears with the same title T1 but a different Resources list, the implementation treats these as distinct, resulting in two tickets with the same title but different resources. 
+
+  
 
-- Each Security Hub Finding type (by title) is represented as a single issue, e.g. if there are three resources that have violated the 'S3.8' rule there will be a single S3.8 Jira issue created
+3. Closes existing Jira issues in the target project if their associated findings are no longer active. For findings with the same title, closure is determined based on the resources’ presence. For example, if there are two tickets for finding T2 with different resources, `r1` and `r2`, and `r1` is no longer present, then only the T2 ticket with `r1` will be closed
```
Comment (yonassrobi) [inline README.md:24]: Is there any performance lag that you noticed when scanning resources that are present on existing tickets?

File: README.md [hunk 3]
```diff
@@ -1,282 +1,584 @@
...
+-  `block`
+
+-  `blocked`
+
+  
 
 For example If there's a workflow which has following transition rules
 
+  
+
```
Comment (yonassrobi) [inline README.md]: This ASCII diagram seems a bit confusing. Any chance you can mock this up on something like diagrams.net and replace it with a screenshot?

File: README.md [hunk 4]
```diff
@@ -1,282 +1,584 @@
...
 **Default Value: ''**
 
+  
+
 **Description:** Comma separated list of User Emails for Atlassion Jira or User EUA IDs for Enterprise Jira.
 
+  
+
 ### `jira-watchers`
```
Comment (yonassrobi) [inline README.md]: `jira-watchers` section is repeated with a different description. Is this intentional?

File: README.md [hunk 5]
```diff
@@ -1,282 +1,584 @@
...
+  
+
+**Description:** a boolean (true/false) to enable/disable the ticket consolidation per run
+  
+  ## test-findings
+**Required: No**  
+
+**Default Value: ''**
+
+**Description:** JSON string of raw findings list
```
Comment (yonassrobi) [inline README.md]: Is this a flag for doing a dry run with dummy data? If so, can it be named something like "dry-run-test-data" and add more details to the description?

File: action.yml [hunk 6]
```diff
@@ -72,6 +72,12 @@ inputs:
   jira-labels-config:
     description: "This variable is to specify stringified labels config"
     required: false
+  jira-consolidate-tickets:
+    description: "True/False value to enable/disable the consolidation of tickets"
+    required: false
+  test-findings:
```
Comment (yonassrobi) [inline action.yml]: same comment as above

File: src/index.ts [hunk 7]
```diff
@@ -151,6 +151,7 @@ async function run(): Promise<void> {
         'JIRA_ADD_LABELS',
         ''
       ),
+      testFindings: getDefaultInputOrEnv('test-findings', 'TEST_FINDINGS', ''),
```
Comment (yonassrobi) [inline src/index.ts]: Similar comment as above - 'testFindings' sounds ambiguous

File: src/macfc-security-hub-sync.ts [hunk 8]
```diff
@@ -65,8 +72,83 @@ export class SecurityHubJiraSync {
...
+          const b = B[j].Id ?? ''
+          same = (a && b && a.includes(b)) as unknown as boolean
+        }
+        isSimilar = isSimilar && same
+      }
+      return isSimilar
+    }
+    return false
+  }
+  inAlreadyInNew(finding: SecurityHubFinding, List: SecurityHubFinding[]) {
```
Comment (yonassrobi) [inline src/macfc-security-hub-sync.ts]: Should this be `isAlreadyInNew`?

File: src/macfc-security-hub-sync.ts [hunk 9]
```diff
@@ -186,6 +374,7 @@ export class SecurityHubJiraSync {
     return updatesForReturn
   }
   makeResourceList(resources: Resource[] | undefined) {
+    console.log('Resources: ', resources?.length)
```
Comment (yonassrobi) [inline src/macfc-security-hub-sync.ts]: Is this console.log necessary when it is not in debug mode?

File: src/macfc-security-hub-sync.ts [hunk 10]
```diff
@@ -428,6 +631,7 @@ export class SecurityHubJiraSync {
     }
     let newIssueInfo
     try {
+      console.log(newIssueData)
```
Comment (yonassrobi) [inline src/macfc-security-hub-sync.ts]: Is this console.log necessary when it is not in debug mode?

File: src/macfc-security-hub-sync.ts [hunk 11]
```diff
@@ -441,6 +645,7 @@ export class SecurityHubJiraSync {
         )
       }
     } catch (e: unknown) {
+      console.log(JSON.stringify(finding))
```
Comment (yonassrobi) [inline src/macfc-security-hub-sync.ts]: Is this console.log necessary when it is not in debug mode?

File: src/macfc-security-hub-sync.ts [hunk 12]
```diff
@@ -451,30 +656,58 @@ export class SecurityHubJiraSync {
...
+      if (this.shouldCreateIssue(finding, jiraIssues)) {
+        try {
+          const update = await this.createJiraIssueFromFinding(
+            finding,
+            identifyingLabels
+          )
+          updatesForReturn.push(update)
+        } catch (e) {
+          console.log(e)
+          console.log('Moving forward with next findings')
```
Comment (yonassrobi) [inline src/macfc-security-hub-sync.ts]: Is it possible to consolidate the two console.log statements above?

**Review by yonassrobi — COMMENTED** [review]
Thanks a bunch for putting a detailed implementation plan for thoroughly testing the enhancements in this PR.

My comments/questions are added mainly on the documentation section of the PR. Additionally, here are a few more broad questions:

* Were you able to test the Nessus consolidation logic with Enterprise Jira? The evidence I reviewed was for Atlassian Jira.
* Does the Nessus consolidation logic impact Security Hub findings or other findings?
* Did you test if other default products (apart from Nessus) can be skipped without issues?

**Comment by maira-samtek** [comment]
@yonassrobi 
> * **Were you able to test the Nessus consolidation logic with Enterprise Jira?** Yes - created this new doc and listed the results specifically for Ent Jira. https://docs.google.com/document/d/1I6LfNETWbwCrFy-rtcckI4wy9voIOmZHglf82u0zFYw/edit?tab=t.0#heading=h.4s5t9t9npum6
> * **Does the Nessus consolidation logic impact Security Hub findings or other findings?** Consolidation is for all products that have mult findings with the same name/title. It will not impact unless var is set to True, once it will consolidate all products that have mult findings with the same title not only Nessus. 
> * **Did you test if other default products (apart from Nessus) can be skipped without issues?** I will we looking further into this, great catch. TBExplored

File: README.md (hunk 1 above)
Comment (maira-samtek) [reply README.md]: > **make sense to rename "security-hub," to "Security Hub titl**
So in this paragraph we are referring to the Label to identify every issue that is created by this automation.

Comment (maira-samtek) [reply README.md]: > > **make sense to rename "security-hub," to "Security Hub titl**
> > So in this paragraph we are referring to the Label to identify every issue that is created by this automation.

I understand you point - ill update the readme to specify "labels" as the identifiers

Comment (maira-samtek) [reply README.md]: @yonassrobi 

> **Can you provide an example of "custom label configuration" here?**
Example: `jira-labels-config:` `"[{\"labelField\":\"ProductName\",\"labelPrefix\":\"product\",\"labelDelimiter\":\":\"},{\"labelField\":\"severity\"},{\"labelField\":\"accountId\",\"labelDelimiter\":\":\",\"labelPrefix\":\"account\"},{\"labelField\":\"region\"}]"`

Result of the custom labeling above the labels on the issues created will be: "product:Tenable" "High" "account:abc123" "us-east-1"

_Ill add an example on the readme too_

> **It's good that you included the "Note" about modifying custom labels here. Can this be bolded/italicized and on a different line to make sure it's clearly visible?**

Yes Ill do that.

File: README.md (hunk 2 above)
Comment (maira-samtek) [reply README.md:24]: > Is there any performance lag that you noticed when scanning resources that are present on existing tickets?

I ran both versions for 3 times and took average time
For new enhancements: Avg Time is 1m34s for 36 issues - [test runs](https://github.com/Enterprise-CMCS/mac-fc-embedded/actions/runs/11862817467)
For old version: Avg Time is 1m30s for 36 issues - [test runs](https://github.com/Enterprise-CMCS/mac-fc-embedded/actions/runs/11862844136)
we can say that the time is not changed significantly

File: README.md (hunk 4 above)
Comment (maira-samtek) [reply README.md]: This is fixed - jira-labels-config was replaced by jira-watchers in second occurrence

File: README.md (hunk 3 above)
Comment (maira-samtek) [reply README.md]: Yes, replaced with screenshot

File: src/macfc-security-hub-sync.ts (hunk 8 above)
Comment (maira-samtek) [reply src/macfc-security-hub-sync.ts]: Good Catch - Renamed the function to "isAlreadyInNew"

File: src/macfc-security-hub-sync.ts (hunk 12 above)
Comment (maira-samtek) [reply src/macfc-security-hub-sync.ts]: Done!

File: README.md (hunk 5 above)
Comment (maira-samtek) [reply README.md]: Renamed the flag

File: src/macfc-security-hub-sync.ts (hunk 9 above)
Comment (maira-samtek) [reply src/macfc-security-hub-sync.ts]: consolidated

File: src/macfc-security-hub-sync.ts (hunk 10 above)
Comment (maira-samtek) [reply src/macfc-security-hub-sync.ts]: removed

File: src/macfc-security-hub-sync.ts (hunk 11 above)
Comment (maira-samtek) [reply src/macfc-security-hub-sync.ts]: removed

File: action.yml (hunk 6 above)
Comment (maira-samtek) [reply action.yml]: renamed and updated description

File: src/index.ts (hunk 7 above)
Comment (maira-samtek) [reply src/index.ts]: renamed

**Review by yonassrobi — APPROVED** [review]
LGTM. Thanks a ton for all your hard work on this.

**Pull request merged**